database/pipeline.go

```diff
@@ -0,0 +1,253 @@
+// Copyright (c) 2022 Target Brands, Inc. All rights reserved.
+//
+// Use of this source code is governed by the LICENSE file in this repository.
+
+package database
+
+import (
+	"bytes"
+	"compress/zlib"
+	"database/sql"
+	"errors"
+	"io/ioutil"
+
+	"github.com/go-vela/types/library"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	// ErrEmptyPipelineNumber defines the error type when a
+	// Pipeline type has an empty Number field provided.
+	ErrEmptyPipelineNumber = errors.New("empty pipeline number provided")
+
+	// ErrEmptyPipelineRef defines the error type when a
+	// Pipeline type has an empty Ref field provided.
+	ErrEmptyPipelineRef = errors.New("empty pipeline ref provided")
+
+	// ErrEmptyPipelineRepoID defines the error type when a
+	// Pipeline type has an empty RepoID field provided.
+	ErrEmptyPipelineRepoID = errors.New("empty pipeline repo_id provided")
+
+	// ErrEmptyPipelineVersion defines the error type when a
+	// Pipeline type has an empty Version field provided.
+	ErrEmptyPipelineVersion = errors.New("empty pipeline version provided")
+)
+
+// Pipeline is the database representation of a pipeline.
+type Pipeline struct {
+	ID              sql.NullInt64  `sql:"id"`
+	RepoID          sql.NullInt64  `sql:"repo_id"`
+	Number          sql.NullInt32  `sql:"number"`
+	Flavor          sql.NullString `sql:"flavor"`
+	Platform        sql.NullString `sql:"platform"`
+	Ref             sql.NullString `sql:"ref"`
```

Comment:
Should this be `ref` or `commit`?

Initially, I went with `ref` because that's what we call it in the [go-vela/server](https://github.com/go-vela/server/blob/master/api/pipeline.go#L425-L426) codebase:

```go
	// capture query parameters
	ref := ctx.DefaultQuery("ref", repo.GetBranch())
```

And then we pass that `ref` to [the `source.ConfigBackoff()` function](https://github.com/go-vela/server/blob/master/api/pipeline.go#L434-L435):

```go
	// send API call to capture the pipeline configuration file
	config, err := scm.FromContext(ctx).ConfigBackoff(user, repo, ref)
```

However, in other parts of the codebase we pass [a commit SHA to that function](https://github.com/go-vela/server/blob/master/api/webhook.go#L376-L377):

```go
	// send API call to capture the pipeline configuration file
	config, err := scm.FromContext(c).ConfigBackoff(u, r, b.GetCommit())
```

Feels like we should be using `commit` instead of `ref` in the existing code for pipelines?

In the `Build{}` type we store both fields so we could do the same for `Pipeline{}` if there is value?

Reply:
Will pipelines be created during a `deployment` event?

Reply:
Yes, a `pipeline` will be created in the database for every webhook that gets triggered.

The exception to this is any conditions we already have setup that prevent a `build` from being created today.

i.e. https://go-vela.github.io/docs/usage/skipping_build/

Reply:
I think having both fields could provide some level of value. However, I would lean towards `commit` if we only want to store one of them.

Reply:
This is done 👍